python/ray/data/preprocessor.py

```diff
@@ -226,59 +241,88 @@ def _transform(self, dataset: Dataset) -> Dataset:
+                f"Unsupported Dataset format: '{dataset_format}'. Only 'pandas' "
+                "and 'arrow' Dataset formats are supported."
             )
 
         transform_type = self._determine_transform_to_use(dataset_format)
 
+        # Our user facing batch format should only be pandas or numpy, other
+        # formats {arrow, simple} are internal.
         if transform_type == "pandas":
             return dataset.map_batches(self._transform_pandas, batch_format="pandas")
-        elif transform_type == "arrow":
-            return dataset.map_batches(self._transform_arrow, batch_format="pyarrow")
+        elif transform_type == "numpy":
+            return dataset.map_batches(self._transform_numpy, batch_format="numpy")
         else:
             raise ValueError(
                 "Invalid transform type returned from _determine_transform_to_use; "
-                f'"pandas" and "arrow" allowed, but got: {transform_type}'
+                f'"pandas" and "numpy" allowed, but got: {transform_type}'
             )
 
-    def _transform_batch(self, df: "DataBatchType") -> "DataBatchType":
+    def _transform_batch(self, data: "DataBatchType") -> "DataBatchType":
+        # For minimal install to locally import air modules
         import pandas as pd
+        from ray.air.constants import TENSOR_COLUMN_NAME
+        from ray.air.util.data_batch_conversion import _convert_batch_type_to_numpy
 
         try:
             import pyarrow
         except ImportError:
             pyarrow = None
 
-        if isinstance(df, pd.DataFrame):
+        if isinstance(data, pd.DataFrame):
             data_format = "pandas"
-        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+        elif pyarrow is not None and isinstance(data, pyarrow.Table):
             data_format = "arrow"
+        elif isinstance(data, (dict, np.ndarray)):
+            data_format = "numpy"
         else:
             raise NotImplementedError(
-                "`transform_batch` is currently only implemented for Pandas DataFrames "
-                f"and PyArrow Tables. Got {type(df)}."
+                "`transform_batch` is currently only implemented for Pandas "
+                "DataFrames, pyarrow Tables, NumPy ndarray and dictionary of "
+                f"ndarray. Got {type(data)}."
             )
 
         transform_type = self._determine_transform_to_use(data_format)
 
         if transform_type == "pandas":
             if data_format == "pandas":
-                return self._transform_pandas(df)
-            else:
-                return self._transform_pandas(df.to_pandas())
-        elif transform_type == "arrow":
-            if data_format == "arrow":
-                return self._transform_arrow(df)
+                return self._transform_pandas(data)
             else:
-                return self._transform_arrow(pyarrow.Table.from_pandas(df))
+                return self._transform_pandas(data.to_pandas())
```

Comment:
let's use `convert_batch_to_pandas` instead of calling `to_pandas` directly?

Reply:
Actually the logic in this section seems unnecessary? 

we explicitly say in the `transform_batch` docstring that the output type of `transform_batch` may be different than the input type
```
This may differ
                from the input type depending on which ``_transform_*`` method(s)
                are implemented.
```
is this behavior still fine or do we need to change this?


If we keep this same behavrio, then I think this can be simplified down to:
```python
if transform_type == "pandas":
    return self._transform_pandas(_convert_batch_type_to_pandas(data))
elif transform_type == "numpy":
    return self._transform_numpy(_convert_batch_type_to_numpy(data))
```

Reply:
I think this is true for pandas that we can just use the same util function to convert batch into pandas; Slightly more work is needed for numpy tho since both arrow & pandas format prefers returning single column table.